### PR TITLE
Fixes #3121. Store observers in registry in `Set` instead of `List`.

### LINKED ISSUE #3121: ObserverRegistry can store more than one reference to the same observer.
### Problem

[ObserverRegistry](https://github.com/mozilla-mobile/android-components/blob/51ef74db39a20fc76d62775db0f2039a6d0959bd/components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt#L19) stores registered observers in `List`. This may lead to situation when one observer can be subscribed multiple times but only first entry will be removed on [unregister request](https://github.com/mozilla-mobile/android-components/blob/51ef74db39a20fc76d62775db0f2039a6d0959bd/components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt#L65).

### Fix

Using `Set` instead. :)

### PR DESCRIPTION
**Pull request by dector**
### Pull Request checklist
- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] ~~**Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one~~
- [x] ~~**Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features~~

### CODE REVIEW

File: components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt [hunk 1]
```diff
@@ -117,6 +130,7 @@ class ObserverRegistry<T> : Observable<T> {
     /**
      * GenericLifecycleObserver implementation to bind an observer to a Lifecycle.
      */
+    @SuppressLint("RestrictedApi")
```
Comment (NotWoods) [inline components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Is this suppress needed? I can't replicate the lint error.

Comment (dector) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Probably yes. :) `GenericLifecycleObserver` (the one we are extending here) has restricted scope [LIBRARY](https://android.googlesource.com/platform/frameworks/support/+/refs/heads/master/lifecycle/common/src/main/java/androidx/lifecycle/GenericLifecycleObserver.java#25).

Comment (pocmo) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: We should not suppress this and fix that - Let's file a follow-up issue. :)

Comment (NotWoods) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: Follow-up: #3140

Comment (dector) [reply components/support/base/src/main/java/mozilla/components/support/base/observer/ObserverRegistry.kt]: @pocmo @NotWoods Thanks. :+1:

**Pull request merged**